linkerd/router/src/lib.rs

```diff
@@ -0,0 +1,143 @@
+#![deny(rust_2018_idioms, clippy::disallowed_methods, clippy::disallowed_types)]
+#![forbid(unsafe_code)]
+
+use futures::{future, prelude::*};
+use linkerd_error::Error;
+use linkerd_stack::{layer, ExtractParam, NewService, Oneshot, Service, ServiceExt};
+use std::{
+    fmt::Debug,
+    marker::PhantomData,
+    task::{Context, Poll},
+};
+use tracing::debug;
+
+mod cache;
+
+pub use self::cache::{Cache, NewCache};
+
+pub trait SelectRoute<Req> {
+    type Key;
+    type Error: Into<Error>;
+
+    /// Given a a request, returns the key matching this request.
+    ///
+    /// If no route matches the request, this method returns an error.
+    fn select(&self, req: &Req) -> Result<Self::Key, Self::Error>;
+}
+
+/// A [`NewService`] that builds `Route` services for targets that implement
+/// [`SelectRoute`].
+#[derive(Debug)]
+pub struct NewOneshotRoute<Sel, X, N> {
+    extract: X,
+    inner: N,
+    _marker: PhantomData<fn() -> Sel>,
+}
+
+/// Dispatches requests to a new `S`-typed inner service.
+///
+/// Each request is matched against the route table and routed to a new inner
+/// service via a [`Oneshot`].
+#[derive(Clone, Debug)]
+pub struct OneshotRoute<Sel, N> {
+    select: Sel,
+    new_route: N,
+}
+
+// === impl NewOneshotRoute ===
+
+impl<Sel, X: Clone, N> NewOneshotRoute<Sel, X, N> {
+    pub fn new(extract: X, inner: N) -> Self {
+        Self {
+            extract,
+            inner,
+            _marker: PhantomData,
+        }
+    }
+
+    pub fn layer(extract: X) -> impl layer::Layer<N, Service = Self> + Clone {
+        layer::mk(move |inner| Self::new(extract.clone(), inner))
+    }
+
+    /// Returns a new `NewOneshotRoute` layer that retains & reuses its inner services.
+    pub fn layer_cached<K>(
+        extract: X,
+    ) -> impl layer::Layer<N, Service = NewOneshotRoute<Sel, X, NewCache<K, N>>> + Clone {
+        layer::mk(move |inner: N| NewOneshotRoute::new(extract.clone(), NewCache::new(inner)))
+    }
+}
+
+impl<T, Sel, X, N> NewService<T> for NewOneshotRoute<Sel, X, N>
+where
+    X: ExtractParam<Sel, T>,
```

Comment:
what is the difference between using `ExtractParam` here and just requiring that `T: Param<Sel>`? my guess is that it's to allow disambiguating when the target is `Param` of multiple types that implement `SelectRoute`?

should we add a comment explaining what the `ExtractParam` impl is for?

Reply:
i think ideally we wouldn't need extractparam but there are some places now where we use general target types (like `http::Accept`) where we don't want to implement `SelectRoute` on that target type. ideally, we'd use different target types, but that's a bigger change i think.

`ExtractParam` lets us use either approach -- either by providing an extractor or by passing `()` to use `T: Param<Sel>`

Reply:
actually that may not be entirely true... I'll take a closer look

Reply:
simplified the interface without changing the extract behavior -- the current uses would require more target mapping than i'd like. The extract stuff is more general than Param and can describe that behavior as well